fms2_io/include/domain_write.inc

```diff
@@ -1258,4 +1279,575 @@ subroutine domain_write_5d(fileobj, variable_name, vdata, unlim_dim_level, &
+                                                              !! in each dimension.
+
+  integer :: xdim_index
+  integer :: ydim_index
+  integer :: xpos
+  integer :: ypos
+
+  integer :: isd !< Starting "x" index of the data domain
+  integer :: isc !< Starting "x" index of the compute domain
+  integer :: jsd !< Starting "y" index of the data domain
+  integer :: jsc !< Starting "y" index of the compute domain
+  integer :: xc_size !< Size of the compute domain
+  integer :: yc_size !< Size of the compute domain
+  logical :: buffer_includes_halos !< True if "vdata" is the size of the data domain
+                                   !! (i.e it includes halos (which are not written))
+
+  integer, dimension(3) :: c !< Indices of the corners for each dimension
+  integer, dimension(3) :: e !< Size of the data for each dimension
+  integer :: varid
+  integer :: unlim_dim_index
+
+  integer :: i0, i1
+  integer :: j0, j1
+  integer :: err
+
+  if (.not. is_variable_domain_decomposed(fileobj, variable_name, .true., xdim_index, ydim_index, xpos, ypos)) then
+    call compressed_write(fileobj, variable_name, vdata, unlim_dim_level=unlim_dim_level, corner=corner, &
+                          edge_lengths=edge_lengths)
+    return
+  endif
+
+  !! Get some more info about the variable
+  call domain_offsets(size(vdata, xdim_index), size(vdata, ydim_index), fileobj%domain, &
+                      xpos, ypos, isd, isc, xc_size, jsd, jsc, yc_size, &
+                      buffer_includes_halos, msg="file:"//trim(fileobj%path)//" and variable:"//trim(variable_name))
+
+  c = 1
+  e = 1
+
+  c(xdim_index) = isc
+  c(ydim_index) = jsc
+  e(xdim_index) = xc_size
+  e(ydim_index) = yc_size
+
+  if (present(unlim_dim_level)) then
+      unlim_dim_index = get_variable_unlimited_dimension_index(fileobj, variable_name, broadcast=.false.)
+      if (unlim_dim_index .ne. 3) then
+        call error("unlimited dimension must be the slowest varying dimension in variable: "//trim(variable_name))
+      endif
+      c(unlim_dim_index) = unlim_dim_level
+  endif
+
+  i0 = 1
+  j0 = 1
+  select case (xdim_index)
+    case (1)
+      ! X is the first dimension
+      if (buffer_includes_halos) then
+        i0 = isc - isd + 1
+        j0 = jsc - jsd + 1
+      endif
+      i1 = i0 + xc_size - 1
+      j1 = j0 + yc_size - 1
+    case (2)
+      ! X is the second dimension
+      if (buffer_includes_halos) then
+        i0 = jsc - jsd + 1
+        j0 = isc - isd + 1
+      endif
+      i1 = i0 + yc_size - 1
+      j1 = j0 + xc_size - 1
+  end select
+
+  varid = get_variable_id(fileobj%ncid, trim(variable_name), &
+    msg="file:"//trim(fileobj%path)//" and variable:"//trim(variable_name))
+
+  err= nf90_var_par_access(fileobj%ncid, varid, nf90_collective)
+  call check_netcdf_code(err, "Failed to set the variable '"//trim(variable_name)//"' to collective-write mode")
```

Comment:
Instead of doing this here, let's remove this and do this at the end of `netcdf_add_variable`
```   if (fileobj%use_netcdf_mpi) then
    err= nf90_var_par_access(fileobj%ncid, varid, nf90_collective)
    call check_netcdf_code(err, append_error_msg)
  endif 
```